App_Python.tex

```diff
@@ -0,0 +1,79 @@
+
+
+%%%%%%%%%%%
+\section{Datatype Definitions}
+
+With only a few exceptions, the C-based \ac{PMIx} datatypes defined in \chapterref{chap:struct} directly translate to Python. However, Python lacks the size-specific value definitions of C (e.g., \code{uint8_t}) and thus some care must be taken to protect against overflow/underflow situations when moving between the languages. Table\ref{app:python:ctopy} lists the correspondence between datatypes in the two languages.
+
+\begin{landscape}
+    \fontsize{10}{12}
+    \begin{longtable}{ | p{4.5cm} | p{4cm} | p{2cm} | p{6.5cm} |}
+    \caption{C-to-Python Datatype Correspondence} \label{app:python:ctopy} \\
+    \hline
+    C-Definition & C-Name & Python-Name & Python Definition\\ \hline
+    \endhead
+    \code{bool} & PMIX_BOOL & 'bool' & Python \emph{boolean} flag \\ \hline
+    \code{byte} & PMIX_BYTE & 'byte' & A single byte value \\ \hline
+    \code{char*} & PMIX_STRING & 'string' & Either a Python \emph{string} \textbf{OR} a Python \emph{byte array} - if interfacing to a C-based library, these must be converted to the corresponding \code{NULL}-terminated character array \\ \hline
+    \code{size_t} & PMIX_SIZE & 'size' & Integer \\ \hline
+    \code{pid_t} & PMIX_PID & 'pid' & Integer - if interfacing to a C-based library, care must be taken to limit the value to the \code{uint32_t} range \\ \hline
+    \code{int, int8_t, int16_t, int32_t, int64_t} & PMIX_INT, PMIX_INT8, PMIX_INT16, PMIX_INT32, PMIX_INT64 & 'int', 'int8', 'int16', 'int32', 'int64' & Integer - if interfacing to a C-based library, care must be taken to limit the value to its corresponding range \\ \hline
+    \code{uint, uint8_t, uint16_t, uint32_t} & PMIX_UINT, PMIX_UINT8, PMIX_UINT16, PMIX_UINT32, PMIX_UINT64 & 'uint', 'uint8', 'uint16', 'uint32', 'uint64' & Integer - if interfacing to a C-based library, care must be taken to limit the value to its corresponding range \\ \hline
+    \code{float, double} & PMIX_FLOAT, PMIX_DOUBLE & 'float', 'double' & float - if interfacing to a C-based library, care must be taken to limit the value to its corresponding range \\ \hline
+    \code{struct timeval} & PMIX_TIMEVAL & 'timeval' & Either a Python \emph{tuple} containing (sec, microsec) where each field is an integer value, \textbf{OR} a single integer value for the seconds field where the microsecond field shall be assumed to be zero \\ \hline
+    \code{time_t} & PMIX_TIME & 'time' & Must be positive integer \\ \hline
+    \refstruct{pmix_data_type_t} & PMIX_DATA_TYPE & 'dtype' & Integer - if interfacing to a C-based library, care must be taken to limit the value to a \code{uint16_t} range \\ \hline
+    \refstruct{pmix_status_t} & PMIX_STATUS & 'status' & Integer  \\ \hline
```

Comment:
Is the Python name (in this case `status`) namespaced something like `PMIx.status`?
I suspect so, but it's not clear.

Are all of the Python names part of the PMIx namespace or are some native to Python?

Reply:
I need to add some text to better explain it as you have misunderstood what I mean by "Python name". In contrast to C, there is no reason to define new Python datatypes - in fact, it makes programming much harder if you do so. Instead, what I'm showing in the table is what you need to put into (for example) a 'value' object so we know what datatype is intended. Thus, if you have a 'value' object of ('status', 1), we know that we should interpret the integer in that object as being a pmix_status_t when converting to a C library.

Reply:
Ok. I think I understand what you are doing now. Maybe a small example would help in the text. Something like: here is the C snippet and here is the same Python snippet with datatype X

Maybe with a `PMIx_Get` of a job level item - nothing too involved just enough to illustrate.

Reply:
I didn't see an example for this in the appendix, but maybe I overlooked it. A clarifying example would be helpful.